ethereum/sidecar/attestation_validator.go

```diff
@@ -0,0 +1,170 @@
+package sidecar
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"time"
+
+	"cosmossdk.io/log"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	ethconfig "github.com/keep-network/keep-common/pkg/chain/ethereum"
+	ethconnect "github.com/mezo-org/mezod/ethereum"
+	"github.com/mezo-org/mezod/ethereum/bindings/portal"
+)
+
+var defaultIsValidTickerDuration = 30 * time.Second
+
+type attestationValidator struct {
+	logger         log.Logger
+	bridgeContract ethconnect.BridgeContract
+	address        common.Address
+}
+
+func newAttestationValidation(
+	logger log.Logger,
+	bridgeContract ethconnect.BridgeContract,
+	validatorAddress common.Address,
+) *attestationValidator {
+	return &attestationValidator{
+		logger:         logger,
+		bridgeContract: bridgeContract,
+		address:        validatorAddress,
+	}
+}
+
+func (av *attestationValidator) IsValid(ctx context.Context, bridgeAssetsUnlocked *portal.MezoBridgeAssetsUnlocked) (bool, error) {
+	// ticker is used to retry the validation in case
+	// of network transient failure.
+	ticker := time.NewTicker(defaultIsValidTickerDuration)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			ok, err := av.bridgeContract.ValidateAssetsUnlocked(*bridgeAssetsUnlocked)
```

Comment:
`ValidateAssetsUnlocked` returns `true` on valid errors but reverts on invalid ones. We will never have `ok == false` here. That said, retrying here is a no go as we will fall into an infinite loop for invalid entries or entries [that were already confirmed](https://github.com/thesis/mezo-portal/blob/7153bf737ff1c29b8b9a6fbcd96a3fcc1d70662f/solidity/contracts/MezoBridge.sol#L512). I think we should just continue to the next if `IsValid` returns an error. Let's also swap the order of checks in the loop and check `IsConfirmed` first, then `IsValid`.

Reply:
6a676ba6